style: apply /style-guide pass to models/tables (#2725)
wandb/docs: models/tables/visualize-tables.mdx

title: Visualize and analyze tables
---

Customize your W&B Tables to answer questions about your machine learning model's performance, analyze your data, and more. This page is for machine learning practitioners who want to inspect logged data interactively in the W&B App.

Interactively explore your data to:

> * [Compare two W&B Tables](#table-comparison-options) logged as artifact versions to analyze changes in your data or model performance.

Reviewer: Too many reiterations of W&B after the first mention

Author: Dropped the redundant "W&B" (now "Compare two Tables") and trimmed the other repeats on the page.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

* Understand higher-level patterns in your data.
* [View how values you log to a table change throughout your runs](#visualize-how-values-change-throughout-your-runs).

> <Note>

Reviewer: Not sure why this is a callout

Author: Converted the callout to regular prose.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

title: Visualize and analyze tables
---

> Customize your W&B Tables to answer questions about your machine learning model's performance, analyze your data, and more. This page is for machine learning practitioners who want to inspect logged data interactively in W&B.

Reviewer 2: Markety

```suggestion
Customize your W&B Tables to answer questions about your machine learning model's performance, analyze your data, and more.
```

Interactively explore your data to:

> * [Compare two Tables](#table-comparison-options) logged as artifact versions to analyze changes in your data or model performance.

Reviewer 2: ```suggestion
* [Compare two W&B Tables](#table-comparison-options) logged as artifact versions to analyze changes in your data or model performance.
```

* Understand higher-level patterns in your data.
* [View how values you log to a table change throughout your runs](#visualize-how-values-change-throughout-your-runs).

> Tables have the following behaviors:

Reviewer 2: ```suggestion
W&B Tables have the following behaviors:
```

* **Reference each table explicitly**: Use `0`, `1`, and `*` in the filter expression to explicitly specify a column in one or both table instances.
* **Visualize detailed numerical differences as histograms**: Compare the values in any cell at a glance.

### Side-by-side view

To view the two tables side-by-side, change the first dropdown from **Merge Tables: Table** to **List of: Table** and then update the **Page size**. Here the first table selected is on the left and the second one is on the right. You can also compare these tables vertically by clicking the **Vertical** checkbox.

<Frame>
    <img src="/images/data_vis/side_by_side.png" alt="Side-by-side table view" max-width="90%"  />
</Frame>

From the side-by-side view, you can:

* **Compare the tables at a glance**: Apply any operations (sort, filter, group) to both tables in tandem to spot changes or differences. For example, view the incorrect predictions grouped by guess, the hardest negatives overall, or the confidence score distribution by true label.
* **Explore two tables independently**: Scroll through and focus on the side or rows of interest.

## Compare artifacts

> Compare two W&B Tables logged as artifact versions to analyze changes in your data or model performance. Use the merged view or side-by-side view to compare the tables. The following sections describe two common comparison scenarios: comparing the same model over time, and comparing different model variants.

Reviewer 2: ```suggestion
Compare two W&B Tables logged as artifact versions to analyze changes in your data or model performance. Use the merged view or side-by-side view to inspect differences between tables. The following sections describe two common scenarios: comparing the same model over time and comparing different model variants.
```

    <img src="/images/data_vis/merged_view.png" alt="Merged view" max-width="90%"  />
</Frame>

From the merged view, you can:

* **Choose the join key**: Use the dropdown at the top left to set the column to use as the join key for the two tables. Typically this is the unique identifier of each row, such as the filename of a specific example in your dataset or an incrementing index on your generated samples. You can select any column, but some choices may yield illegible tables and slow queries.
* **Concatenate instead of join**: Select "concatenating all tables" in this dropdown to union all the rows from both tables into one larger table instead of joining across their columns.
* **Reference each table explicitly**: Use `0`, `1`, and `*` in the filter expression to explicitly specify a column in one or both table instances.
* **Visualize detailed numerical differences as histograms**: Compare the values in any cell at a glance.

### Side-by-side view

> {/* To do */}

Reviewer: ?

Author: Removed the stray comment here, the matching one under Merged view, and a leftover commented-out table stub.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

6. Set **Render As** selector to **Stepper**.
7. Set **Stepper Key** to `_step` or the [key to use as the unit](#custom-step-keys) for the step slider.

> 

Author: Good catch — fixed in 7910ed9. The heading is `### Custom step key` (singular), so the anchor is `#custom-step-key`. Updated the link to match.

Now, if the slider is set to `epoch = 1`, the following happens:

* `vague-cloud` finds `epoch = 1` and returns the value logged at `_step = 5`.
* `lively-frog` finds `epoch = 1` and returns the value logged at `_step = 4`.
* `serene-sponge` finds `epoch = 1` and returns the value logged at `_step = 2`.

If the slider is set to `epoch = 9`:

* `vague-cloud` also doesn't log `epoch = 9`, so W&B uses the latest prior value `epoch = 3` and returns the value logged at `_step = 20`.
* `lively-frog` doesn't log `epoch = 9`, but the latest prior value is `epoch = 5` so it returns the value logged at `_step = 20`.
* `serene-sponge` finds `epoch = 9` and returns the value logged at `_step = 18`.

## Save your view

> This section describes how table state is preserved across sessions and how to export a table snapshot to a report. Tables you interact with in the run workspace, project workspace, or a report automatically save their view state. If you apply any table operations then close your browser, the table retains the last viewed configuration when you next navigate to the table. 

Reviewer 2: ```suggestion
This section describes how W&B preserves table state across sessions and how to export a table snapshot to a report. Tables in a run workspace, project workspace, or report automatically save their view state. When you apply table operations and return to the table later, the table restores your last viewed configuration.
```

## Examples

> The following reports highlight the different use cases of W&B Tables and show the techniques on this page applied to real datasets:

Reviewer 2: This is an assertion that the bot didn't verify. Don't add.
```suggestion
The following reports highlight the different use cases of W&B Tables:
```
